### PR TITLE
[SQSERVICES 1953] Support OAuth Auhorization Code Flow with PKCE

### PR DESCRIPTION
**Pull request by battermann**
Change the OAuth flow from authorization code flow to authorization code flow with PKCE.

After this PR only PKCE is supported. For more info, see https://wearezeta.atlassian.net/browse/SQSERVICES-1953

## Checklist

 - [x] Add a new entry in an appropriate subdirectory of `changelog.d`
 - [x] Read and follow the [PR guidelines](https://docs.wire.com/developer/developer/pr-guidelines.html)

### CODE REVIEW

File: services/brig/src/Brig/API/OAuth.hs [hunk 1]
```diff
@@ -314,31 +300,24 @@ lookupOauthClient cid = do
...
+insertOAuthAuthorizationCode :: (MonadClient m) => Word64 -> OAuthAuthorizationCode -> OAuthClientId -> UserId -> OAuthScopes -> RedirectUrl -> OAuthCodeChallenge -> m ()
+insertOAuthAuthorizationCode ttl code cid uid scope uri chal = do
   let cqlScope = C.Set (Set.toList (unOAuthScopes scope))
-  retry x5 . write q $ params LocalQuorum (code, cid, uid, cqlScope, uri)
+  retry x5 . write q $ params LocalQuorum (code, cid, uid, cqlScope, uri, chal)
   where
-    q :: PrepQuery W (OAuthAuthorizationCode, OAuthClientId, UserId, C.Set OAuthScope, RedirectUrl) ()
-    q = fromString $ "INSERT INTO oauth_auth_code (code, client, user, scope, redirect_uri) VALUES (?, ?, ?, ?, ?) USING TTL " <> show ttl
+    q :: PrepQuery W (OAuthAuthorizationCode, OAuthClientId, UserId, C.Set OAuthScope, RedirectUrl, OAuthCodeChallenge) ()
+    q = fromString $ "INSERT INTO oauth_auth_code (code, client, user, scope, redirect_uri, code_challenge) VALUES (?, ?, ?, ?, ?, ?) USING TTL " <> show ttl
```
Comment (stefanwire) [inline services/brig/src/Brig/API/OAuth.hs:309]: I know it's from before, but why is `ttl` not part of the query parameters? It is everywhere else in the code.

File: services/brig/src/Brig/API/OAuth.hs [hunk 2]
```diff
@@ -124,7 +125,7 @@ data CreateNewOAuthCodeError
   | CreateNewOAuthCodeErrorRedirectUrlMissMatch
 
 validateAndCreateAuthorizationCode :: UserId -> CreateOAuthAuthorizationCodeRequest -> ExceptT CreateNewOAuthCodeError (Handler r) OAuthAuthorizationCode
-validateAndCreateAuthorizationCode uid (CreateOAuthAuthorizationCodeRequest cid scope responseType redirectUrl _) = do
+validateAndCreateAuthorizationCode uid (CreateOAuthAuthorizationCodeRequest cid scope responseType redirectUrl _state _ chal) = do
```
[the remark this answers is not in the document]
Comment (battermann) [reply services/brig/src/Brig/API/OAuth.hs:128]: Do you mean a `state` parameter? This is implemented. It's just not needed in this function. It's added as query parameter to the redirect URL in one of the functions above.

[the remark this answers is not in the document]
Comment (flokli) [reply services/brig/src/Brig/API/OAuth.hs:128]: Yes, i meant `state`, sorry.

File: services/brig/src/Brig/API/OAuth.hs [hunk 3]
```diff
@@ -175,14 +173,14 @@ verifyRefreshToken key rt = do
 createAccessTokenWithAuthorizationCode :: (Member Now r, Member Jwk r) => OAuthAccessTokenRequest -> (Handler r) OAuthAccessTokenResponse
 createAccessTokenWithAuthorizationCode req = do
   unless (req.grantType == OAuthGrantTypeAuthorizationCode) $ throwStd $ errorToWai @'OAuthInvalidGrantType
-  (cid, uid, scope, uri) <-
-    lift (wrapClient $ lookupAndDeleteByOAuthAuthorizationCode req.code)
+  (cid, uid, scope, uri, mChal) <-
+    lift (wrapClient $ lookupAndDeleteByOAuthAuthorizationCode (req.code))
```
Comment (stefanwire) [inline services/brig/src/Brig/API/OAuth.hs:177]: Interesting, why do we need the parentheses around `req.code`?

File: services/brig/src/Brig/API/OAuth.hs [hunk 4]
```diff
@@ -348,23 +327,23 @@ lookupAndDeleteByOAuthAuthorizationCode code = lookupOAuthAuthorizationCode <* d
...
-    qInsertId = fromString $ "INSERT INTO oauth_user_refresh_token (user, token_id) VALUES (?, ?) USING TTL " <> show ttl
+    qInsertId :: PrepQuery W (UserId, OAuthRefreshTokenId, Int32) ()
+    qInsertId = fromString $ "INSERT INTO oauth_user_refresh_token (user, token_id) VALUES (?, ?) USING TTL ?"
 
     determineOldestTokensToBeDeleted :: [OAuthRefreshTokenInfo] -> [OAuthRefreshTokenInfo]
     determineOldestTokensToBeDeleted tokens =
       take (length sorted - fromIntegral maxActiveTokens + 1) sorted
       where
-        sorted = sortOn createdAt tokens
+        sorted = sortOn (.createdAt) tokens
```
Comment (stefanwire) [inline services/brig/src/Brig/API/OAuth.hs:346]: I would buy the record syntax everywhere else, but here it seems redundant. It's a matter of taste, for sure.

**Review by stefanwire — APPROVED** [review]

File: services/brig/src/Brig/API/OAuth.hs (hunk 3 above)
Comment (battermann) [reply services/brig/src/Brig/API/OAuth.hs:177]: We don't. Was probably due to a merge issue.

File: services/brig/src/Brig/API/OAuth.hs (hunk 4 above)
Comment (battermann) [reply services/brig/src/Brig/API/OAuth.hs:346]: Yes, I agree. It would be needed if this was an overloaded field. But it is not, so I'll remove it.

**Pull request merged**